WIP: Updated shift model and API

Theodlz opened the pull request:
### WIP: This PR is an updated version of the shift model and API. Why is it needed ?

#### Issues with the current model : 

- Main issue : Shifts belong to a group, but there is no way for this group's users to join, or not join a shift. So we can't assign people to shifts, we can't see who will participate in a shift.
- Secondary issue: We can't add any description/details about a shift. So unless you make it very clear in its name, it is impossible to know what a shift is about / what is its goal (we won't go in details here, we simply added a column to the shift table to add the description, and it is displayed in the frontend).

What we need, is to allow users the group to which the shift belongs, to join or leave that particular shift. And of course, in the DB, there is something that allows you to create a group of users, that can ask to join, join and leave it: the groups.


#### Solution: Instead of belonging to one "main" group, a shift is now linked to 2 distinct groups.

-  The first one is still the **"main group"**. A shift belongs to it, and only members from that main group can see that shift in the shift page and interact with it.
- The second one, is a unique group created at the same time as the shift. This **"shift group"**, is what allows us to have users participate or not to a specific shift. On the frontend, users will be able in the next PR to join or leave that shift group. Now, as users can participate to a shift, or not, we are also able to display the list of people participating to a shift.

#### What has been implemented yet (we will update that list when we make progress):

1. The models for the shifts and the groups has been modified (to add new columns and update relationships as required by sqlalchemy).
2. The handler for the shift has been modified (to create the associated "shift group" automatically when we create a shift).
3. The frontend elements that were in conflict with the backend changes. We made minimal changes here, but we are planning on adding the buttons to join and leave a shift).

#### What to do next ?

We are currently working on another PR that aims to update the frontend to add a calendar element to the Shift Page. You can find it here: [WIP: Updated shift frontend using new backend and API #2701](https://github.com/skyportal/skyportal/pull/2701)

Theodlz commented:
Hi @mcoughlin , tagging you here to let you know we opened that backend PR.

profjsb commented on skyportal/handlers/api/shift.py:
```diff
@@ -42,14 +38,47 @@ def post(self):
         except ValidationError as exc:
             return self.error(
-                f"Invalid/missing parameters: {exc.normalized_messages()}"
+                f"2Invalid/missing parameters: {exc.normalized_messages()}"
```
2?

profjsb commented on skyportal/models/group.py:
```diff
@@ -302,11 +302,19 @@ class Group(Base):
+        sa.Boolean,
+        default=False,
+        index=True,
+        doc='Flag indicating whether this group ' 'is a shift group or not.',
```
```suggestion
        doc='Flag indicating whether this group is a shift group or not.',
```

profjsb commented on skyportal/tests/api/test_shifts.py:
```diff
@@ -1,11 +1,30 @@
 def test_shift(public_group, super_admin_token, view_only_token, super_admin_user):
 
+    shift_name = str(uuid.uuid4())
+    group_description = "Group description"
```
```suggestion
    group_description = "Shift Group description"
```

profjsb commented on static/js/components/NewShift.jsx:
```diff
@@ -62,6 +62,10 @@ const NewShift = () => {
       },
+      name: {
+        type: "string",
+        title: "Shift name (i.e. the Night Shift)",
```
```suggestion
        title: "Shift name (ie. the Night Shift)",
```

profjsb reviewed (changes requested):
Makes sense to make -- thanks for the context in the comments. Either you'll need to modify the group administrative pages as well as any page that shows groups (like classifications, comments, etc.) to only show non-shift groups OR you could create a new model class ShiftGroup. I strongly prefer the latter since you wont need to tough a lot of other code. In this case you'd create a subclass of Group:

```python
class ShiftGroup(Group):
```

which would take as it's cascade delete foreign key one single group ID. Make sense?

Theodlz commented:
Hi @profjsb , thank you so much for your review and suggestions. You're right, it does make sense to create a subclass of the group model, which is basically creating a new table in the DB. I was also thinking, if we create a new table with this subclass, maybe we can also use that table to hold all the shift infos (start date, end date, ...), so to have one unique table for the shifts instead of having 2 of them ? I'll work on adding that subclass then. Have a nice day !

Theodlz commented:
> Makes sense to me -- thanks for the context in the comments. Either you'll need to modify the group administrative pages as well as any page that shows groups (like classifications, comments, etc.) to only show non-shift groups OR you could create a new model class ShiftGroup. I strongly prefer the latter since you wont need to tough a lot of other code. In this case you'd create a subclass of Group:
> 
> ```python
> class ShiftGroup(Group):
> ```
> 
> which would take as it's cascade delete foreign key one single group ID. Make sense?

Hi @profjsb , regarding this class ShiftGroup inheriting from Group, what is the expected behaviour you have in mind ? What i had in mind and what i understood from your comment is the following: the ShiftGroup class inherits the same columns as a Group, so it has a name, a description, relationships with users and GroupUsers, ..., and to that we add 2 columns related to shifts: start_date and end_date. But when i use something like : 
 ```python
 class ShiftGroup(Group):
> ```

It does not do that, I doesn't not inherit the columns of the Group table, but sql achemy asks me to have one of the columns of ShiftGroup to be a foreign key to a column in the Group Table. 

Am I doing it wrong or did I not understand what you meant ?

profjsb commented on skyportal/handlers/api/shift.py:
```diff
@@ -145,3 +174,247 @@ def delete(self, shift_id):
+
+
+class ShiftUserHandler(BaseHandler):
+    @permissions(["Upload data"])
```
should be a new role "Manage shifts"

profjsb reviewed (changes requested):
Looking very good. Some small changes suggested. Please also add some data as appropriate to the demo data. Also, will need some new tests.

Theodlz commented on skyportal/handlers/api/shift.py:
```diff
@@ -145,3 +174,247 @@ def delete(self, shift_id):
+                    type: boolean
+                  canSave:
+                    type: boolean
+                    description: Boolean indicating whether user can save sources to shift. Defaults to true.
```
Hi @profjsb , it's a simple copy and paste issue. As mentionned it is a Work in Progress and we copied a lot from the groups table to rapidly get something up and working. We'll delete this soon.

Theodlz commented on skyportal/handlers/api/shift.py:
```diff
@@ -145,3 +174,247 @@ def delete(self, shift_id):
+            return self.error("Invalid userID parameter: unable to parse to integer")
+
+        admin = data.get("admin", False)
+        if not isinstance(admin, bool):
```
You're right. By the way this comes from the group handler, so maybe it needs to be corrected in the group handler too.

Theodlz commented on skyportal/handlers/api/shift.py:
```diff
@@ -145,3 +174,247 @@ def delete(self, shift_id):
+        return self.success()
+
+    @auth_or_token
+    def delete(self, shift_id, user_id):
```
We are actually hesitating. We want to allow a user of a group to join a shift and leave it freely. So really the only thing to check (that we considered), is that the user belongs to the group that has the shift. We could also verify the permission to have a "difference" between users that can participate to shifts or not in a specific group. We'll think about it.

Theodlz commented on static/js/components/NewShift.jsx:
```diff
@@ -74,12 +78,12 @@ const NewShift = () => {
       },
     },
-    required: ["group_id", "start_date", "end_date"],
+    required: ["group_id", "name", "start_date", "end_date"],
```
II don't think so. We were thinking that giving a name to a shift should be mandatory, but the description optional. What do you think ?

Theodlz commented on skyportal/handlers/api/shift.py:
```diff
@@ -41,7 +45,23 @@ def post(self):
+
         data = self.get_json()
+        try:
+            shift_admin_ids = [int(e) for e in data.get('shift_admins', [])]
```
Sure. I'll do it

Theodlz commented:
> Looking very good. Some small changes suggested. Please also add some data as appropriate to the demo data. Also, will need some new tests.

Hi Josh, only thing missing now is the demo data and the tests, but otherwise we addressed your comments the best we can, and fixed other "bugs" that are more behaviors that didn't meet our requirements.

Theodlz commented:
Hi @guynir42 @profjsb, could you have a look at the new changes please ? Thanks for your help.

guynir42 commented on skyportal/tests/api/test_shifts.py:
```diff
@@ -8,17 +8,18 @@ def test_shift(public_group, super_admin_token, view_only_token, super_admin_use
-    status, data = api('GET', f'shift/{public_group.id}', token=super_admin_token)
+    status, data = api('GET', f'shifts/{public_group.id}', token=super_admin_token)
     assert status == 200
     assert data['status'] == 'success'
```
Maybe check that the shift you posted is in fact recovered. I'd add a uuid name for the shift and make sure you can find that name among one of the returned shifts.

Theodlz replied:
That's a great idea, i'll do that now

guynir42 commented on static/js/ducks/shifts.js:
```diff
@@ -8,8 +8,34 @@ const FETCH_SHIFTS_OK = "skyportal/FETCH_SHIFTS_OK";
+  API.PATCH(`/api/shifts/${shiftID}/users`, UPDATE_SHIFT_USER, params);
+
+export function deleteShiftUser({ userID, shift_id }) {
+  console.log(userID, shift_id);
```
no need for console.log.

Theodlz replied:
Yep. Forgot to delete it earlier

guynir42 reviewed (commented):
I have minor comments. This looks very close to going in.

profjsb reviewed (commented):
Looking good (i agree with @guynir42's comments). Will take a look after @guynir42 approves.

Theodlz commented on skyportal/handlers/api/shift.py:
```diff
@@ -53,6 +103,9 @@ def post(self):
 
         DBSession().add(shift)
+        DBSession().add_all(
+            [ShiftUser(shift=shift, user=user, admin=True) for user in shift_admins]
```
It works in the same way as the groups. When creating a group, the current user will be added as its admin, unless you specific an array of sfhit_admins. Later on, using the ShiftUserHandler, you can post non admin users to add them to a shift.

Theodlz commented on skyportal/handlers/api/shift.py:
```diff
@@ -36,13 +51,48 @@ def post(self):
+            .filter(User.id.in_(shift_admin_ids))
+            .all()
+        )
+        if self.current_user not in shift_admins and not isinstance(
```
You're right. I mean the shift_admins list is a list of user IDs after all, so thats what should be tested. I think the same changes should be applied in the GroupHandler

Theodlz replied:
I think you mentioned this issue last week. From what i saw, it is not just for the shifts but also for the groups and other handlers. Should we just decide a general / common way to handle that in all handlers ?

guynir42 commented on skyportal/tests/api/test_shifts.py:
```diff
@@ -1,13 +1,20 @@
+    name = str(uuid.uuid4())
+    start_date = date.today().strftime("%Y-%m-%dT%H:%M:%S")
+    end_date = (date.today() + timedelta(days=1)).strftime("%Y-%m-%dT%H:%M:%S")
+    print(start_date)
+    print(end_date)
```
```suggestion
```

guynir42 commented on skyportal/tests/api/test_shifts.py:
```diff
@@ -23,6 +30,10 @@ def test_shift(public_group, super_admin_token, view_only_token, super_admin_use
-
     assert any([request_data['end_date'] == s['end_date'] for s in data["data"]])
+    assert (
+        len([s for s in data['data'][0]['users'] if s['id'] == super_admin_user.id])
```
Can't guarantee that your shift will be at index 0. would probably need to do two loops, one with `if s['name']==name` and one inside that to check the users.

Theodlz replied:
Oh yes you're very right i tried that first and forgot to add the 2nd loop. Will do it now

guynir42 reviewed (commented):
Looks good. Two minor comments.

Theodlz commented:
> @Theodlz also please re-run any failing tests to make sure the failures are unrelated.

Hey @guynir42 , i keep getting an error related to the migrations. I'll redo that part entirely soon.
